{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2b57ce43",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr22.hap2.tsv\n",
      "h2tg000436l: Strand not consistent between the first and last segments.\n",
      "h2tg000633l+: Gap from 45175 to end on chr22:88600-unknown\n",
      "h2tg000170l+: Gap from 15451 to end on chr22:184817-unknown\n",
      "h2tg000385l+: Gap from 165230 to 219716 on chr22:2047332-2100445\n",
      "h2tg000385l+: Gap from 306823 to end on chr22:2183879-unknown\n",
      "h2tg000788l-: Gap from 167233 to end on chr22:unknown-1949629\n",
      "h2tg000356l+: Gap from 106610 to 523841 on chr22:2635652-2992651\n",
      "h2tg000356l+: Gap from 553148 to 872852 on chr22:3021389-2990912\n",
      "h2tg000356l+: Gap from 895466 to 969137 on chr22:3013478-3145482\n",
      "h2tg000356l+: Gap from 1007956 to 1079746 on chr22:3185297-3264565\n",
      "h2tg000303l-: Gap from 392174 to 805085 on chr22:3189752-3081873\n",
      "h2tg000303l-: Gap from 974844 to 1230486 on chr22:3074409-3079592\n",
      "h2tg000303l-: Gap from 1295742 to 1367034 on chr22:3143654-2999312\n",
      "h2tg000303l+: Gap from 1436478 to 2453676 on chr22:2995511-8432367\n",
      "h2tg000303l+: Gap from 2480263 to end on chr22:8457748-unknown\n",
      "h2tg000635l+: Gap from 64624 to end on chr22:4178309-unknown\n",
      "h2tg003187l+: Gap from 17064 to end on chr22:4452211-unknown\n",
      "h2tg000236l+: Gap from 1370916 to 1484428 on chr22:6905459-7151808\n",
      "h2tg000236l-: Gap from 1500328 to 1709309 on chr22:7170816-9734503\n",
      "h2tg000236l: Strand not consistent between the first and last segments.\n",
      "h2tg000082l-: Gap from 15856 to end on chr22:unknown-6211865\n",
      "h2tg000148l-: Gap from 3358031 to 3520376 on chr22:10741460-9509066\n",
      "h2tg000148l-: Gap from 3889547 to 5297559 on chr22:9558290-7559560\n",
      "h2tg000148l: Strand not consistent between the first and last segments.\n",
      "h2tg000260l+: Gap from 3218257 to 3272324 on chr22:9693409-10410397\n",
      "h2tg000260l+: Gap from 3582129 to end on chr22:10717911-unknown\n",
      "h2tg000254l+: Gap from 194075 to 270163 on chr22:13720791-13158993\n",
      "h2tg000254l+: Gap from 718240 to 790938 on chr22:14268821-13895527\n",
      "h2tg000072l-: Gap from 874689 to 1936066 on chr22:15085880-15441180\n",
      "h2tg000072l-: Gap from 1961435 to end on chr22:unknown-15441180\n",
      "h2tg000217l+: Gap from 793810 to 1024009 on chr22:16465743-16418907\n",
      "h2tg001828l: Strand not consistent between the first and last segments.\n",
      "h2tg000196l: Strand not consistent between the first and last segments.\n",
      "h2tg000428l+: Gap from 24032 to end on chr22:23713418-unknown\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr11.hap1.tsv\n",
      "h1tg000612l+: Gap from 192922 to end on chr11:192482-unknown\n",
      "h1tg000486l-: Gap from 143255 to 193556 on chr11:52316280-52462511\n",
      "h1tg000113l+: Gap from 1118125 to 1170420 on chr11:53176087-53984868\n",
      "h1tg001083l-: Gap from 60641 to 131733 on chr11:51732182-53226632\n",
      "h1tg002791l-: Gap from 40904 to end on chr11:unknown-52198574\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr9.hap2.tsv\n",
      "h2tg000097l+: Gap from 20971029 to end on chr9:72321655-unknown\n",
      "h2tg000143l-: Gap from 892627 to 943883 on chr9:62425242-54771812\n",
      "h2tg000143l: Strand not consistent between the first and last segments.\n",
      "h2tg000143l-: Gap from 1093157 to 1156978 on chr9:53413074-62144457\n",
      "h2tg000143l-: Gap from 1197588 to 1258275 on chr9:75300543-53054133\n",
      "h2tg000143l: Strand not consistent between the first and last segments.\n",
      "h2tg000143l+: Gap from 2291370 to 2396353 on chr9:60923243-58300002\n",
      "h2tg000143l: Strand not consistent between the first and last segments.\n",
      "h2tg000143l+: Gap from 4198788 to 4276712 on chr9:64553916-73012855\n",
      "h2tg000143l+: Gap from 4814033 to 4878324 on chr9:52506119-62256437\n",
      "h2tg000143l-: Gap from 6971308 to 7039674 on chr9:43434070-47573534\n",
      "h2tg000143l-: Gap from 10886158 to 10978063 on chr9:46229723-45495642\n",
      "h2tg000529l+: Gap from 68853 to 164055 on chr9:73096577-56578181\n",
      "h2tg000529l+: Gap from 318865 to end on chr9:65150652-unknown\n",
      "h2tg000294l: Strand not consistent between the first and last segments.\n",
      "h2tg000135l+: Gap from 7419718 to end on chr9:150577184-unknown\n",
      "h2tg000099l-: Gap from 220825 to end on chr9:unknown-150399448\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr6.hap2.tsv\n",
      "h2tg000034l+: Gap from 21205247 to 21376256 on chr6:59168746-59724816\n",
      "h2tg001069l: Strand not consistent between the first and last segments.\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr21.hap1.tsv\n",
      "h1tg000045l+: Gap from 270499 to 336032 on chr21:680192-801024\n",
      "h1tg000045l+: Gap from 701497 to 975576 on chr21:1269039-2013234\n",
      "h1tg000858l+: Gap from 116487 to end on chr21:6424277-unknown\n",
      "h1tg000219l-: Gap from 741852 to end on chr21:unknown-6810932\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap2.tsv\n",
      "h2tg000283l+: Gap from 1157809 to 1857323 on chr12:35601616-35703107\n",
      "h2tg000090l-: Gap from 4651173 to 4729880 on chr12:36136698-35561322\n",
      "h2tg000064l-: Gap from 1401495 to 1453922 on chr12:131694379-123453035\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr5.hap1.tsv\n",
      "h1tg005813l: Strand not consistent between the first and last segments.\n",
      "h1tg000462l-: Gap from 52079 to 109612 on chr5:49607125-49539915\n",
      "h1tg000462l-: Gap from 188537 to 408144 on chr5:49549272-49528382\n",
      "h1tg000462l-: Gap from 479949 to 656308 on chr5:49456136-49339838\n",
      "h1tg000462l-: Gap from 807923 to 1099133 on chr5:49475168-49422560\n",
      "h1tg000462l-: Gap from 1175368 to 1235356 on chr5:49471925-49416263\n",
      "h1tg000583l-: Gap from 136696 to 370314 on chr5:49471921-49426433\n",
      "h1tg000374l-: Gap from 1206179 to 1256787 on chr5:51919734-49994386\n",
      "h1tg000374l-: Gap from 2251894 to 2305081 on chr5:49887079-49767268\n",
      "h1tg000374l-: Gap from 2482919 to 2550294 on chr5:49714353-49636089\n",
      "h1tg000374l-: Gap from 2651748 to end on chr5:unknown-49602417\n",
      "h1tg000147l+: Gap from 839288 to end on chr5:170488125-unknown\n",
      "h1tg000110l-: Gap from 139075 to end on chr5:unknown-181896680\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr16.hap1.tsv\n",
      "h1tg000405l+: Gap from 1612885 to end on chr16:35764732-unknown\n",
      "h1tg000200l-: Gap from 831295 to 919206 on chr16:40041718-39950591\n",
      "h1tg000200l: Strand not consistent between the first and last segments.\n",
      "h1tg000295l: Strand not consistent between the first and last segments.\n",
      "h1tg000248l+: Gap from 402202 to 472125 on chr16:41048238-41066859\n",
      "h1tg000248l+: Gap from 1263234 to 1330900 on chr16:48018768-48075770\n",
      "h1tg000248l+: Gap from 1626114 to 1759953 on chr16:48303820-48444482\n",
      "h1tg000248l+: Gap from 1784047 to 1902884 on chr16:48474746-48548943\n",
      "h1tg000170l-: Gap from 175732 to end on chr16:unknown-96210429\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr19.hap1.tsv\n",
      "h1tg000170l+: Gap from 82042 to 134758 on chr19:79795-152444\n",
      "h1tg000986l-: Gap from 39146 to 114457 on chr19:29501230-29352231\n",
      "h1tg000986l-: Gap from 140453 to 319989 on chr19:29378219-29003549\n",
      "h1tg002496c+: Gap from 16248 to end on chr19:29094819-unknown\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr1.hap2.tsv\n",
      "h2tg000228l+: Gap from 2192268 to 2290078 on chr1:2133067-2133223\n",
      "h2tg000271l: Strand not consistent between the first and last segments.\n",
      "h2tg001175l: Strand not consistent between the first and last segments.\n",
      "h2tg000339l: Strand not consistent between the first and last segments.\n",
      "h2tg000339l-: Gap from 87077 to end on chr1:unknown-54606592\n",
      "h2tg000086l+: Gap from 250931 to 339346 on chr1:129696670-135121320\n",
      "h2tg000086l+: Gap from 684004 to 739124 on chr1:129504078-140525472\n",
      "h2tg000086l+: Gap from 1473644 to 1523904 on chr1:132849340-133204120\n",
      "h2tg000086l+: Gap from 1559827 to 1618546 on chr1:130403456-130877044\n",
      "h2tg000086l+: Gap from 1724081 to 1775593 on chr1:141619380-130343174\n",
      "h2tg000086l+: Gap from 1891945 to 1956757 on chr1:137829894-134496910\n",
      "h2tg000086l-: Gap from 3632932 to 3691689 on chr1:125012009-125991766\n",
      "h2tg000086l-: Gap from 3856372 to 3911994 on chr1:124191513-126016783\n",
      "h2tg000086l: Strand not consistent between the first and last segments.\n",
      "h2tg000086l-: Gap from 4984831 to 5102793 on chr1:126025898-126090682\n",
      "h2tg000086l+: Gap from 5530439 to 5607583 on chr1:124332865-125057551\n",
      "h2tg000086l: Strand not consistent between the first and last segments.\n",
      "h2tg000086l+: Gap from 6191125 to 6260749 on chr1:125467529-125079658\n",
      "h2tg000086l+: Gap from 6520187 to 6590020 on chr1:124889788-125192213\n",
      "h2tg000086l-: Gap from 8250736 to 8362428 on chr1:125931715-122116610\n",
      "h2tg000086l: Strand not consistent between the first and last segments.\n",
      "h2tg000086l+: Gap from 8416923 to 8474237 on chr1:121982027-124490537\n",
      "h2tg000086l: Strand not consistent between the first and last segments.\n",
      "h2tg000109l-: Gap from 18738663 to 19195485 on chr1:124021659-125212830\n",
      "h2tg000109l-: Gap from 19417763 to 19471184 on chr1:122107524-125312180\n",
      "h2tg000109l: Strand not consistent between the first and last segments.\n",
      "h2tg000109l-: Gap from 19499358 to 19678769 on chr1:125340344-125002370\n",
      "h2tg000109l+: Gap from 19709855 to 19830737 on chr1:125033454-121799437\n",
      "h2tg000109l+: Gap from 19948906 to 20022517 on chr1:124318907-123334203\n",
      "h2tg000109l: Strand not consistent between the first and last segments.\n",
      "h2tg000109l-: Gap from 20155014 to 20263981 on chr1:122156646-124788044\n",
      "h2tg000109l-: Gap from 20425669 to 20535626 on chr1:122329856-124130516\n",
      "h2tg000109l: Strand not consistent between the first and last segments.\n",
      "h2tg000109l-: Gap from 20558471 to 20619928 on chr1:124153343-125405331\n",
      "h2tg000109l-: Gap from 20646372 to 20755834 on chr1:125431766-124221727\n",
      "h2tg000109l+: Gap from 20788799 to 20964649 on chr1:124254678-123481849\n",
      "h2tg000109l-: Gap from 20994612 to 21049269 on chr1:123511818-124780832\n",
      "h2tg000109l+: Gap from 21161073 to 21295048 on chr1:122210546-123103724\n",
      "h2tg000109l: Strand not consistent between the first and last segments.\n",
      "h2tg000109l+: Gap from 21328727 to 21396819 on chr1:123137410-122157571\n",
      "h2tg000109l+: Gap from 21576729 to 21716491 on chr1:124502439-122040267\n",
      "h2tg000109l: Strand not consistent between the first and last segments.\n",
      "h2tg000109l: Strand not consistent between the first and last segments.\n",
      "h2tg000249l+: Gap from 136378 to 189807 on chr1:121990093-121883836\n",
      "h2tg000249l+: Gap from 1304908 to 1379755 on chr1:123790810-123799699\n",
      "h2tg000249l+: Gap from 2406345 to 2457876 on chr1:125993883-124701338\n",
      "h2tg000249l: Strand not consistent between the first and last segments.\n",
      "h2tg000249l+: Gap from 2862551 to 2935222 on chr1:125371140-125285458\n",
      "h2tg000249l+: Gap from 4201731 to 4257904 on chr1:126968253-126998876\n",
      "h2tg000249l: Strand not consistent between the first and last segments.\n",
      "h2tg000040l-: Gap from 205439 to 442698 on chr1:122037398-121976520\n",
      "h2tg000040l-: Gap from 470467 to 903546 on chr1:122004438-122933080\n",
      "h2tg000040l-: Gap from 930642 to end on chr1:unknown-122933080\n",
      "h2tg000112l+: Gap from 1595057 to end on chr1:122002623-unknown\n",
      "h2tg001188l: Strand not consistent between the first and last segments.\n",
      "h2tg000188l-: Gap from 89071 to 245700 on chr1:129684623-130420618\n",
      "h2tg000188l-: Gap from 266540 to 574996 on chr1:130442532-140178032\n",
      "h2tg000188l-: Gap from 621655 to 703914 on chr1:136090340-129698757\n",
      "h2tg000188l-: Gap from 805658 to 856841 on chr1:140521118-140490611\n",
      "h2tg000188l+: Gap from 919548 to 1055257 on chr1:141991340-126936749\n",
      "h2tg000188l+: Gap from 1301439 to end on chr1:127069658-unknown\n",
      "h2tg000260l+: Gap from 359080 to 456684 on chr1:139788775-127436489\n",
      "h2tg000260l-: Gap from 721967 to 827869 on chr1:127000280-134653176\n",
      "h2tg000260l-: Gap from 1817577 to 3664794 on chr1:129058828-127553014\n",
      "h2tg000260l: Strand not consistent between the first and last segments.\n",
      "h2tg000209l-: Gap from 382758 to 536771 on chr1:127359949-135121323\n",
      "h2tg000209l: Strand not consistent between the first and last segments.\n",
      "h2tg000209l-: Gap from 616299 to 696454 on chr1:129730288-140928645\n",
      "h2tg000209l-: Gap from 730754 to 819133 on chr1:140963159-129662805\n",
      "h2tg000209l-: Gap from 895021 to 1009210 on chr1:127408562-129464090\n",
      "h2tg000209l: Strand not consistent between the first and last segments.\n",
      "h2tg000209l+: Gap from 1329066 to 1383231 on chr1:127167555-127013110\n",
      "h2tg000209l: Strand not consistent between the first and last segments.\n",
      "h2tg000209l-: Gap from 1598954 to 1650909 on chr1:129721719-133083445\n",
      "h2tg000209l: Strand not consistent between the first and last segments.\n",
      "h2tg000209l-: Gap from 1675629 to 1735598 on chr1:133107645-129654243\n",
      "h2tg000209l-: Gap from 1835204 to end on chr1:unknown-129666465\n",
      "h2tg000320l+: Gap from 71502 to 176326 on chr1:142111026-140551285\n",
      "h2tg000320l-: Gap from 419614 to 625271 on chr1:141336481-127074097\n",
      "h2tg001313l-: Gap from 37343 to 117055 on chr1:127225062-142150487\n",
      "h2tg001313l+: Gap from 137707 to 201617 on chr1:142171823-127197571\n",
      "h2tg000142l-: Gap from 184497 to 260657 on chr1:127225062-142150487\n",
      "h2tg000142l+: Gap from 281309 to 345219 on chr1:142171823-127197571\n",
      "h2tg002555l: Strand not consistent between the first and last segments.\n",
      "h2tg000218l-: Gap from 52732 to 138293 on chr1:136443387-140354730\n",
      "h2tg000218l+: Gap from 240874 to 355256 on chr1:131723183-127413538\n",
      "h2tg000218l: Strand not consistent between the first and last segments.\n",
      "h2tg000218l-: Gap from 719243 to end on chr1:unknown-138847370\n",
      "h2tg000111l+: Gap from 405073 to 884741 on chr1:141250690-127435184\n",
      "h2tg000111l+: Gap from 906997 to end on chr1:127457790-unknown\n",
      "h2tg000148l+: Gap from 3612789 to 4624084 on chr1:128027210-128346749\n",
      "h2tg000148l+: Gap from 4689671 to 4962022 on chr1:128410436-128668787\n",
      "h2tg000148l+: Gap from 5001202 to end on chr1:128718485-unknown\n",
      "h2tg000335l+: Gap from 483503 to end on chr1:128412423-unknown\n",
      "h2tg000145l-: Gap from 166314 to 250354 on chr1:129308297-130334138\n",
      "h2tg000145l-: Gap from 308050 to 369451 on chr1:139629789-129696481\n",
      "h2tg000145l-: Gap from 387695 to 599325 on chr1:129714749-130420618\n",
      "h2tg000145l-: Gap from 622513 to 767291 on chr1:130444201-129650189\n",
      "h2tg000261l+: Gap from 73953 to 212895 on chr1:129686619-129666663\n",
      "h2tg000261l+: Gap from 232496 to 318681 on chr1:129686619-129698377\n",
      "h2tg000261l+: Gap from 383816 to 480700 on chr1:134179315-129698708\n",
      "h2tg000261l+: Gap from 506381 to 658133 on chr1:129723849-133083445\n",
      "h2tg000261l+: Gap from 750528 to 889532 on chr1:141703641-129696624\n",
      "h2tg000261l+: Gap from 910073 to end on chr1:129717230-unknown\n",
      "h2tg000149l+: Gap from 89414 to 227872 on chr1:129720781-134936155\n",
      "h2tg000149l+: Gap from 401622 to 556284 on chr1:129696277-129696624\n",
      "h2tg000149l+: Gap from 624232 to 741738 on chr1:140109705-140084779\n",
      "h2tg000149l+: Gap from 774822 to 891965 on chr1:129722211-129698221\n",
      "h2tg000149l+: Gap from 999975 to 1334943 on chr1:129730356-140495880\n",
      "h2tg000149l+: Gap from 1360258 to 1632028 on chr1:140521106-134655723\n",
      "h2tg000149l+: Gap from 1647324 to 1742401 on chr1:134670672-129667075\n",
      "h2tg000149l+: Gap from 1759813 to 1984629 on chr1:129684623-129696921\n",
      "h2tg000370l+: Gap from 253988 to 355846 on chr1:139650327-129666654\n",
      "h2tg000370l: Strand not consistent between the first and last segments.\n",
      "h2tg000370l-: Gap from 539703 to end on chr1:unknown-141974789\n",
      "h2tg000293l-: Gap from 75438 to 181893 on chr1:129685090-137126143\n",
      "h2tg000293l-: Gap from 203223 to 329911 on chr1:137149437-132249035\n",
      "h2tg000293l-: Gap from 349505 to 527262 on chr1:132269018-130932507\n",
      "h2tg000293l-: Gap from 549610 to 734727 on chr1:130954701-130732695\n",
      "h2tg000293l-: Gap from 835251 to 905102 on chr1:137220194-141252760\n",
      "h2tg000293l-: Gap from 940805 to 1036444 on chr1:141288100-138177457\n",
      "h2tg000293l-: Gap from 1049698 to 1228277 on chr1:138189850-136515541\n",
      "h2tg000293l-: Gap from 1245544 to 1701877 on chr1:136533066-133632219\n",
      "h2tg000801l-: Gap from 54895 to end on chr1:unknown-129686406\n",
      "h2tg000274l+: Gap from 201089 to 274417 on chr1:129723795-129695596\n",
      "h2tg000274l+: Gap from 334567 to end on chr1:129723421-unknown\n",
      "h2tg000140l-: Gap from 178075 to end on chr1:unknown-129743030\n",
      "h2tg000242l+: Gap from 69129 to 180547 on chr1:129992798-141590273\n",
      "h2tg000242l+: Gap from 240722 to 482857 on chr1:142108786-129701623\n",
      "h2tg001121c-: Gap from 26896 to 108321 on chr1:129762176-130931907\n",
      "h2tg001121c-: Gap from 130801 to end on chr1:unknown-130931907\n",
      "h2tg000253l+: Gap from 231677 to 284434 on chr1:134839397-129742238\n",
      "h2tg000253l+: Gap from 405502 to 460521 on chr1:140136546-129985287\n",
      "h2tg000253l+: Gap from 888898 to 987209 on chr1:132904712-130186239\n",
      "h2tg000416l-: Gap from 198740 to end on chr1:unknown-130330700\n",
      "h2tg000657l+: Gap from 83439 to end on chr1:130440632-unknown\n",
      "h2tg000722l+: Gap from 41543 to 96167 on chr1:133692805-136539593\n",
      "h2tg000722l+: Gap from 119970 to end on chr1:136562234-unknown\n",
      "h2tg000506l+: Gap from 21473 to 491007 on chr1:137149437-140121345\n",
      "h2tg000506l+: Gap from 511336 to 687864 on chr1:140140013-139675898\n",
      "h2tg000407l-: Gap from 138500 to end on chr1:unknown-137762478\n",
      "h2tg001531l+: Gap from 13450 to end on chr1:140497764-unknown\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chrX.hap2.tsv\n",
      "h2tg000054l-: Gap from 10812945 to 10878023 on chrX:59074451-56999975\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr15.hap2.tsv\n",
      "h2tg000303l-: Gap from 2253450 to 2407551 on chr15:559353-367802\n",
      "h2tg000788l-: Gap from 134460 to end on chr15:unknown-913873\n",
      "h2tg000424l-: Gap from 693876 to 782995 on chr15:1380793-1074092\n",
      "h2tg000284l-: Gap from 217113 to end on chr15:unknown-1732514\n",
      "h2tg000158l+: Gap from 1755705 to 1933360 on chr15:17847684-5414801\n",
      "h2tg000170l-: Gap from 461279 to end on chr15:unknown-5564217\n",
      "h2tg000154l+: Gap from 604103 to 658760 on chr15:11713132-8738369\n",
      "h2tg000154l+: Gap from 828197 to 901798 on chr15:11780276-12372401\n",
      "h2tg000154l+: Gap from 1137414 to 1202005 on chr15:6649837-11399023\n",
      "h2tg000154l+: Gap from 1750147 to 1833571 on chr15:8828899-11287985\n",
      "h2tg000210l+: Gap from 422870 to 473569 on chr15:15770473-14380306\n",
      "h2tg000474l+: Gap from 394921 to 603123 on chr15:17493770-17394968\n",
      "h2tg000474l+: Gap from 689857 to 1022347 on chr15:17475728-16688936\n",
      "h2tg000474l+: Gap from 1067176 to 1225776 on chr15:16733878-16714607\n",
      "h2tg000659l-: Gap from 214580 to 289874 on chr15:16832840-16690073\n",
      "h2tg000371l-: Gap from 123926 to end on chr15:unknown-16709032\n",
      "h2tg000219l-: Gap from 741594 to 913527 on chr15:17198169-16862406\n",
      "h2tg000001l: Strand not consistent between the first and last segments.\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap1.tsv\n",
      "h1tg000828l: Strand not consistent between the first and last segments.\n",
      "h1tg001947l: Strand not consistent between the first and last segments.\n",
      "h1tg003948l: Strand not consistent between the first and last segments.\n",
      "h1tg000840l: Strand not consistent between the first and last segments.\n",
      "h1tg000279l-: Gap from 174924 to 232030 on chr2:93725204-93760786\n",
      "h1tg000279l-: Gap from 819132 to 936616 on chr2:93505327-92854525\n",
      "h1tg000279l-: Gap from 1364935 to 1453598 on chr2:92637082-93701291\n",
      "h1tg000279l-: Gap from 1473549 to 1556205 on chr2:93720902-94190433\n",
      "h1tg000219l-: Gap from 202404 to end on chr2:unknown-91996578\n",
      "h1tg001020l-: Gap from 78684 to end on chr2:unknown-93378298\n",
      "h1tg000007l: Strand not consistent between the first and last segments.\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr20.hap1.tsv\n",
      "h1tg000870l: Strand not consistent between the first and last segments.\n",
      "h1tg000870l+: Gap from 291781 to end on chr20:26413767-unknown\n",
      "h1tg000258l-: Gap from 2759829 to 2842328 on chr20:32209772-32084967\n",
      "h1tg000258l-: Gap from 2881625 to 2935408 on chr20:32113481-31807918\n",
      "h1tg000258l-: Gap from 3254211 to 3305585 on chr20:31638127-31625844\n",
      "h1tg000093l-: Gap from 104469 to end on chr20:unknown-66103776\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr13.hap2.tsv\n",
      "h2tg000356l+: Gap from 842644 to 892865 on chr13:1277325-949047\n",
      "h2tg000356l+: Gap from 1088624 to end on chr13:1247859-unknown\n",
      "h2tg000303l-: Gap from 217843 to 310081 on chr13:871657-1133543\n",
      "h2tg000303l-: Gap from 567240 to 621800 on chr13:948525-830539\n",
      "h2tg000303l-: Gap from 783156 to 909123 on chr13:794902-1105109\n",
      "h2tg000303l-: Gap from 1243947 to 1443276 on chr13:316613-796688\n",
      "h2tg000303l-: Gap from 1758807 to end on chr13:unknown-1686270\n",
      "h2tg000050l-: Gap from 57441 to 184112 on chr13:657414-625994\n",
      "h2tg000050l-: Gap from 312151 to 601460 on chr13:654338-625994\n",
      "h2tg000050l-: Gap from 636520 to 947393 on chr13:662961-635461\n",
      "h2tg000050l-: Gap from 1044369 to 1372781 on chr13:642694-631586\n",
      "h2tg000050l-: Gap from 1408277 to end on chr13:unknown-631586\n",
      "h2tg000241l: Strand not consistent between the first and last segments.\n",
      "h2tg000241l-: Gap from 133619 to end on chr13:unknown-4721944\n",
      "h2tg000666l+: Gap from 96004 to end on chr13:4915073-unknown\n",
      "h2tg000487l-: Gap from 736631 to end on chr13:unknown-9886849\n",
      "h2tg000170l-: Gap from 232003 to 352018 on chr13:10403390-10255425\n",
      "h2tg000170l-: Gap from 372090 to end on chr13:unknown-10255425\n",
      "h2tg000296l-: Gap from 31534 to end on chr13:unknown-10987757\n",
      "h2tg000158l-: Gap from 1714820 to end on chr13:unknown-11017720\n",
      "h2tg000165l+: Gap from 331158 to 430821 on chr13:15500579-15287092\n",
      "h2tg000165l+: Gap from 833563 to 911336 on chr13:15437685-15432996\n",
      "h2tg000058l-: Gap from 8804492 to 10489941 on chr13:25688148-15034667\n",
      "h2tg000626l-: Gap from 20917 to end on chr13:unknown-79776596\n",
      "h2tg000317l: Strand not consistent between the first and last segments.\n",
      "h2tg000739l: Strand not consistent between the first and last segments.\n",
      "h2tg000330l+: Gap from 1220867 to end on chr13:107235115-unknown\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr4.hap1.tsv\n",
      "h1tg000700l-: Gap from 394203 to 500142 on chr4:53876512-52689082\n",
      "h1tg000700l-: Gap from 548580 to end on chr4:unknown-52689082\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr10.hap1.tsv\n",
      "h1tg001168l-: Gap from 34066 to 120408 on chr10:39656553-41800419\n",
      "h1tg000497c+: Gap from 24431 to 100659 on chr10:39656553-39624839\n",
      "h1tg000497c+: Gap from 133069 to end on chr10:39656553-unknown\n",
      "h1tg001186l+: Gap from 29071 to 178380 on chr10:41516066-41039044\n",
      "h1tg001186l+: Gap from 274958 to 345074 on chr10:40149708-41039125\n",
      "h1tg001186l+: Gap from 377500 to 491540 on chr10:41072223-41627622\n",
      "h1tg000117l-: Gap from 307701 to 499797 on chr10:41647720-41006016\n",
      "h1tg000117l-: Gap from 575004 to end on chr10:unknown-41039044\n",
      "h1tg001342l-: Gap from 129324 to end on chr10:unknown-41046934\n",
      "h1tg002252l-: Gap from 45681 to 114646 on chr10:41611160-41627622\n",
      "h1tg000005l+: Gap from 35746887 to end on chr10:86969665-unknown\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr7.hap2.tsv\n",
      "h2tg000161l-: Gap from 10225249 to 10292560 on chr7:62710455-60804532\n",
      "h2tg000161l-: Gap from 11825234 to 11877122 on chr7:60519662-60793895\n",
      "h2tg000161l-: Gap from 12521440 to 12585497 on chr7:60515842-61930496\n",
      "h2tg000161l-: Gap from 13523582 to 13576353 on chr7:61040742-60629479\n",
      "h2tg000496l-: Gap from 783470 to 864077 on chr7:61732202-61432861\n",
      "h2tg000496l-: Gap from 1171607 to 1242548 on chr7:61320148-60907074\n",
      "h2tg000496l-: Gap from 1297179 to 1492848 on chr7:60966132-62040786\n",
      "h2tg000496l-: Gap from 2171338 to 2251974 on chr7:61732202-61432861\n",
      "h2tg000496l-: Gap from 2559504 to 2630445 on chr7:61320148-60907074\n",
      "h2tg000496l-: Gap from 2685076 to 2802736 on chr7:60966132-61535390\n",
      "h2tg000496l-: Gap from 3160686 to end on chr7:unknown-61682564\n",
      "h2tg000062l-: Gap from 987408 to 1083538 on chr7:104887146-93771301\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr8.hap2.tsv\n",
      "h2tg000815l: Strand not consistent between the first and last segments.\n",
      "h2tg000013l+: Gap from 20709050 to 20904076 on chr8:44344747-44770756\n",
      "h2tg000013l+: Gap from 21257238 to 21390282 on chr8:45185333-45336326\n",
      "h2tg000201l: Strand not consistent between the first and last segments.\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr14.hap2.tsv\n",
      "h2tg000385l+: Gap from 346335 to end on chr14:821103-unknown\n",
      "h2tg000284l-: Gap from 11244 to end on chr14:unknown-1662772\n",
      "h2tg000260l-: Gap from 1904441 to 2468539 on chr14:4993966-4324029\n",
      "h2tg000260l-: Gap from 3054984 to end on chr14:unknown-4105524\n",
      "h2tg000072l-: Gap from 1516456 to 1575307 on chr14:11983440-11864716\n",
      "h2tg000148l+: Gap from 4297361 to 4375664 on chr14:4717268-4895855\n",
      "h2tg000148l+: Gap from 4710882 to 5085873 on chr14:5271773-5554530\n",
      "h2tg000128l-: Gap from 1803229 to end on chr14:unknown-6443206\n",
      "h2tg000210l-: Gap from 457874 to end on chr14:unknown-7908842\n",
      "h2tg000083l: Strand not consistent between the first and last segments.\n",
      "h2tg000217l+: Gap from 351953 to 726375 on chr14:13852387-13858818\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr3.hap1.tsv\n",
      "h1tg000626l-: Gap from 166586 to end on chr3:unknown-200895549\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr18.hap1.tsv\n",
      "h1tg000174l: Strand not consistent between the first and last segments.\n",
      "h1tg000652l: Strand not consistent between the first and last segments.\n",
      "h1tg000716l: Strand not consistent between the first and last segments.\n",
      "h1tg000720l-: Gap from 667068 to 747076 on chr18:18255933-17117617\n",
      "h1tg000702l+: Gap from 252371 to 308278 on chr18:20674448-20730086\n",
      "h1tg001004l: Strand not consistent between the first and last segments.\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr17.hap1.tsv\n",
      "h1tg000447l: Strand not consistent between the first and last segments.\n",
      "h1tg002252l+: Gap from 20286 to end on chr17:27570020-unknown\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr3.hap2.tsv\n",
      "h2tg000158l-: Gap from 1812018 to end on chr3:unknown-562337\n",
      "h2tg000241l+: Gap from 201354 to end on chr3:92834578-unknown\n",
      "h2tg000003l-: Gap from 2823784 to 2881307 on chr3:198727899-194314954\n",
      "h2tg000003l: Strand not consistent between the first and last segments.\n",
      "h2tg000003l-: Gap from 6959490 to 7072154 on chr3:198400830-165604681\n",
      "h2tg000003l-: Gap from 105515034 to 105566691 on chr3:95899800-95270513\n",
      "h2tg000003l-: Gap from 108199292 to 108274840 on chr3:93338114-93435857\n",
      "h2tg000003l-: Gap from 109392568 to 109468116 on chr3:93338114-93435857\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr14.hap1.tsv\n",
      "h1tg000040l-: Gap from 67095 to end on chr14:unknown-749731\n",
      "h1tg000219l+: Gap from 1893002 to 2084671 on chr14:4822303-4825538\n",
      "h1tg000219l+: Gap from 2102585 to 2198918 on chr14:4841947-4841749\n",
      "h1tg000219l+: Gap from 3258517 to end on chr14:5894051-unknown\n",
      "h1tg000438l-: Gap from 147243 to end on chr14:unknown-4099375\n",
      "h1tg000846l+: Gap from 993417 to end on chr14:13852387-unknown\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr18.hap2.tsv\n",
      "h2tg000637l: Strand not consistent between the first and last segments.\n",
      "h2tg002125l: Strand not consistent between the first and last segments.\n",
      "h2tg000343l+: Gap from 407402 to end on chr18:18980687-unknown\n",
      "h2tg000410l: Strand not consistent between the first and last segments.\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr17.hap2.tsv\n",
      "h2tg000336l: Strand not consistent between the first and last segments.\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr20.hap2.tsv\n",
      "h2tg000560l: Strand not consistent between the first and last segments.\n",
      "h2tg000296l-: Gap from 48057 to end on chr20:unknown-26222545\n",
      "h2tg000117l+: Gap from 188197 to 266826 on chr20:27005744-28082391\n",
      "h2tg000117l+: Gap from 710136 to 808676 on chr20:27451670-28342464\n",
      "h2tg000117l+: Gap from 1138242 to 1619717 on chr20:27528363-28428829\n",
      "h2tg000056l-: Gap from 33380802 to 33462868 on chr20:66206460-32681750\n",
      "h2tg000056l-: Gap from 35968191 to 36079596 on chr20:28277553-28161711\n",
      "h2tg000210l-: Gap from 535719 to end on chr20:unknown-31616176\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr4.hap2.tsv\n",
      "h2tg000389l: Strand not consistent between the first and last segments.\n",
      "h2tg000170l+: Gap from 168586 to end on chr4:49458204-unknown\n",
      "h2tg000022l-: Gap from 19128629 to 19213551 on chr4:55167574-54987771\n",
      "h2tg000022l-: Gap from 19536051 to 19597629 on chr4:54926175-54910572\n",
      "h2tg000022l-: Gap from 19655270 to end on chr4:unknown-51190095\n",
      "h2tg000542l+: Gap from 246536 to end on chr4:53239719-unknown\n",
      "h2tg000461l-: Gap from 93488 to end on chr4:unknown-52689082\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr13.hap1.tsv\n",
      "h1tg000442l: Strand not consistent between the first and last segments.\n",
      "h1tg003825l: Strand not consistent between the first and last segments.\n",
      "h1tg005709l: Strand not consistent between the first and last segments.\n",
      "h1tg001666l: Strand not consistent between the first and last segments.\n",
      "h1tg000594l: Strand not consistent between the first and last segments.\n",
      "h1tg000594l+: Gap from 560224 to end on chr13:91901373-unknown\n",
      "h1tg001345l+: Gap from 215247 to end on chr13:91901373-unknown\n",
      "h1tg002099l+: Gap from 26802 to end on chr13:91901373-unknown\n",
      "h1tg000261l+: Gap from 175953 to end on chr13:107235115-unknown\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr7.hap1.tsv\n",
      "h1tg000582l: Strand not consistent between the first and last segments.\n",
      "h1tg000054l+: Gap from 76571 to end on chr7:61572016-unknown\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr8.hap1.tsv\n",
      "h1tg000177l-: Gap from 11642126 to end on chr8:unknown-6999439\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr10.hap2.tsv\n",
      "h2tg000155l+: Gap from 103624 to 229865 on chr10:41895913-41644427\n",
      "h2tg000392l+: Gap from 23410 to 114843 on chr10:41505545-41630413\n",
      "h2tg000392l+: Gap from 139911 to 278316 on chr10:41656501-41634064\n",
      "h2tg000392l+: Gap from 453047 to 525343 on chr10:41661708-39682832\n",
      "h2tg000392l+: Gap from 748583 to 866931 on chr10:41100840-41629353\n",
      "h2tg000392l+: Gap from 930485 to end on chr10:40953546-unknown\n",
      "h2tg000121l-: Gap from 171404 to 291778 on chr10:41659485-41046934\n",
      "h2tg000121l-: Gap from 393979 to 583674 on chr10:41072155-41040106\n",
      "h2tg000121l-: Gap from 613518 to 824672 on chr10:41070636-41619805\n",
      "h2tg000121l-: Gap from 852247 to end on chr10:unknown-41619805\n",
      "h2tg000239l-: Gap from 213999 to 280251 on chr10:41592960-41039124\n",
      "h2tg000138l-: Gap from 4771457 to 4850914 on chr10:42591700-42461503\n",
      "h2tg000161l-: Gap from 13903524 to end on chr10:unknown-41877096\n",
      "h2tg000017l-: Gap from 22142 to end on chr10:unknown-134737392\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr1.hap1.tsv\n",
      "h1tg000472l+: Gap from 37346 to 176927 on chr1:124802983-123267368\n",
      "h1tg000472l+: Gap from 208841 to 327886 on chr1:123298638-123853984\n",
      "h1tg000665l: Strand not consistent between the first and last segments.\n",
      "h1tg000665l-: Gap from 50859 to end on chr1:unknown-125600986\n",
      "h1tg001149l-: Gap from 16857 to end on chr1:unknown-126239136\n",
      "h1tg001324l-: Gap from 80945 to end on chr1:unknown-127020554\n",
      "h1tg000614l: Strand not consistent between the first and last segments.\n",
      "h1tg000614l+: Gap from 119725 to end on chr1:129322738-unknown\n",
      "h1tg001615l+: Gap from 10883 to end on chr1:127052234-unknown\n",
      "h1tg000239l: Strand not consistent between the first and last segments.\n",
      "h1tg001204l-: Gap from 172351 to 248511 on chr1:142171823-127197571\n",
      "h1tg001204l: Strand not consistent between the first and last segments.\n",
      "h1tg000140c-: Gap from 82800 to 162512 on chr1:142171823-127197571\n",
      "h1tg000140c-: Gap from 189805 to 348283 on chr1:127225062-127197571\n",
      "h1tg000312l-: Gap from 27397 to 152984 on chr1:129072090-142091760\n",
      "h1tg000312l: Strand not consistent between the first and last segments.\n",
      "h1tg000312l-: Gap from 216896 to 384258 on chr1:140957377-140891212\n",
      "h1tg000312l-: Gap from 406630 to end on chr1:unknown-140891212\n",
      "h1tg001949l+: Gap from 22080 to 226335 on chr1:127483479-133360838\n",
      "h1tg001949l+: Gap from 266124 to end on chr1:133400340-unknown\n",
      "h1tg000438l+: Gap from 790487 to 862455 on chr1:127511515-140899535\n",
      "h1tg000438l+: Gap from 1618354 to 1689815 on chr1:141075544-137732263\n",
      "h1tg000438l+: Gap from 1711022 to end on chr1:137753185-unknown\n",
      "h1tg000219l+: Gap from 38436 to end on chr1:128359603-unknown\n",
      "h1tg000380l+: Gap from 58647 to 407595 on chr1:129072090-129966967\n",
      "h1tg000380l+: Gap from 431877 to end on chr1:129992798-unknown\n",
      "h1tg000264l-: Gap from 166314 to 263582 on chr1:129308297-130334138\n",
      "h1tg000264l-: Gap from 290373 to 384455 on chr1:130361939-129696481\n",
      "h1tg000264l-: Gap from 402699 to 572745 on chr1:129714749-130420618\n",
      "h1tg000264l-: Gap from 664534 to 740711 on chr1:139685914-129650189\n",
      "h1tg001240l-: Gap from 43062 to end on chr1:unknown-129479010\n",
      "h1tg000272l+: Gap from 128509 to 604601 on chr1:129684623-129671288\n",
      "h1tg000807l-: Gap from 146749 to 225081 on chr1:141098950-129666654\n",
      "h1tg000807l-: Gap from 244510 to 346368 on chr1:129686331-139626657\n",
      "h1tg000235l+: Gap from 118098 to 290692 on chr1:136801738-129698767\n",
      "h1tg000235l+: Gap from 310028 to end on chr1:129720462-unknown\n",
      "h1tg000395l-: Gap from 38224 to 129528 on chr1:140142085-140370178\n",
      "h1tg000395l-: Gap from 268391 to 402107 on chr1:138499480-129742238\n",
      "h1tg000395l-: Gap from 420891 to 508939 on chr1:129762176-130931908\n",
      "h1tg000395l-: Gap from 531419 to 934401 on chr1:130954701-141282931\n",
      "h1tg000395l-: Gap from 964757 to 1042853 on chr1:141314484-137805032\n",
      "h1tg000395l-: Gap from 1056107 to 1144380 on chr1:137817425-137282782\n",
      "h1tg000395l-: Gap from 1225778 to end on chr1:unknown-134821285\n",
      "h1tg001579l+: Gap from 34964 to end on chr1:130408987-unknown\n",
      "h1tg000196l-: Gap from 134939 to 385574 on chr1:142172853-140527816\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr16.hap2.tsv\n",
      "h2tg000516l: Strand not consistent between the first and last segments.\n",
      "h2tg000425l-: Gap from 91421 to end on chr16:unknown-36100455\n",
      "h2tg000091l+: Gap from 97354 to 350900 on chr16:36418127-36672755\n",
      "h2tg000109l+: Gap from 20024055 to 20450469 on chr16:37829457-37439737\n",
      "h2tg000109l+: Gap from 20486095 to 21665437 on chr16:37475355-37507007\n",
      "h2tg000109l+: Gap from 21694966 to 21918158 on chr16:37536501-37159431\n",
      "h2tg000109l+: Gap from 21963327 to end on chr16:37459741-unknown\n",
      "h2tg000357l-: Gap from 298002 to 414134 on chr16:40180496-40042424\n",
      "h2tg000357l-: Gap from 450132 to 508014 on chr16:40077001-39983490\n",
      "h2tg000316l+: Gap from 299847 to 373508 on chr16:40097539-41396637\n",
      "h2tg000316l+: Gap from 532732 to 612919 on chr16:39815145-39878676\n",
      "h2tg000316l+: Gap from 1209756 to 1269125 on chr16:49095402-51580432\n",
      "h2tg000316l+: Gap from 1555453 to 1647388 on chr16:50513954-51975011\n",
      "h2tg000181l-: Gap from 978496 to 1035656 on chr16:51125687-51574646\n",
      "h2tg000181l-: Gap from 1432351 to 1493320 on chr16:40016421-39901075\n",
      "h2tg000318l+: Gap from 735627 to 888264 on chr16:48744264-48915741\n",
      "h2tg000318l+: Gap from 1453631 to 1697573 on chr16:44762631-40164351\n",
      "h2tg000085l-: Gap from 1411318 to 1498370 on chr16:49047845-48036414\n",
      "h2tg000085l-: Gap from 1756130 to 1824361 on chr16:41771416-41641072\n",
      "h2tg000085l-: Gap from 1856090 to 1949661 on chr16:41672722-40072212\n",
      "h2tg000085l-: Gap from 2649072 to 2716841 on chr16:46549568-39996135\n",
      "h2tg000169l+: Gap from 164347 to 225690 on chr16:40903757-40969426\n",
      "h2tg000034l-: Gap from 28295 to end on chr16:unknown-84282314\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr19.hap2.tsv\n",
      "h2tg000109l+: Gap from 19029714 to 19159416 on chr19:29506325-29034945\n",
      "h2tg000109l+: Gap from 19181931 to 19270268 on chr19:29057461-25828786\n",
      "h2tg000109l+: Gap from 19649091 to 19725539 on chr19:29501595-29094197\n",
      "h2tg000109l+: Gap from 19783650 to 19867531 on chr19:29042201-29358216\n",
      "h2tg000109l+: Gap from 19896461 to 20091535 on chr19:29387141-26798532\n",
      "h2tg000109l+: Gap from 20279669 to 20334250 on chr19:25848422-29545241\n",
      "h2tg000109l+: Gap from 20355746 to 20424986 on chr19:29566736-29098068\n",
      "h2tg000109l+: Gap from 20446579 to 20546095 on chr19:29119651-29582018\n",
      "h2tg000109l+: Gap from 20733927 to 20794255 on chr19:29506326-29018593\n",
      "h2tg000109l+: Gap from 20817357 to 20884408 on chr19:29041692-28431088\n",
      "h2tg000109l+: Gap from 21109004 to 21179149 on chr19:29037515-29095062\n",
      "h2tg000109l+: Gap from 21203850 to 21256327 on chr19:29119756-29365515\n",
      "h2tg000109l+: Gap from 21341899 to 21458851 on chr19:29506326-26825308\n",
      "h2tg000109l+: Gap from 21551266 to 21619459 on chr19:29125628-28841297\n",
      "h2tg000109l+: Gap from 21664937 to 21740711 on chr19:29562713-29063520\n",
      "h2tg000109l+: Gap from 21783536 to end on chr19:25846656-unknown\n",
      "h2tg000032l-: Gap from 553345 to 605468 on chr19:26621323-26338094\n",
      "h2tg000073l-: Gap from 12625678 to end on chr19:unknown-25818434\n",
      "h2tg000112l+: Gap from 287485 to 345229 on chr19:29557558-29032913\n",
      "h2tg000112l+: Gap from 514589 to 577569 on chr19:29119649-29518957\n",
      "h2tg000112l+: Gap from 1134979 to 1236184 on chr19:29514977-29056887\n",
      "h2tg000112l+: Gap from 2720362 to 2771604 on chr19:29491001-28365956\n",
      "h2tg000086l-: Gap from 3320847 to 3790054 on chr19:28238607-25921055\n",
      "h2tg000086l-: Gap from 3811979 to 8095004 on chr19:25942792-28206025\n",
      "h2tg000086l-: Gap from 8126755 to end on chr19:unknown-28206025\n",
      "h2tg000129l-: Gap from 240556 to end on chr19:unknown-29474186\n",
      "h2tg000249l+: Gap from 2910340 to 3213928 on chr19:29615025-29325500\n",
      "h2tg000249l+: Gap from 3241173 to end on chr19:29352786-unknown\n",
      "h2tg000040l-: Gap from 274297 to end on chr19:unknown-29450916\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/hap2.ctg.multimap.tsv\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 352, in main\n",
      "    seqid, hapid, tmp = get_scaffold(filename, len_thres=len_thres, gap_thres=gap_thres)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 128, in get_scaffold\n",
      "    is_longctg = check_ctg(ctg, prev, len_thres) # aka pass len_thres and not contained\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 45, in check_ctg\n",
      "    elif ctg[1] < len_thres: return False\n",
      "TypeError: '<' not supported between instances of 'str' and 'int'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chrX.hap1.tsv\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr2.hap2.tsv\n",
      "h2tg000386l: Strand not consistent between the first and last segments.\n",
      "h2tg000291l-: Gap from 1510141 to 1561337 on chr2:92891216-92488338\n",
      "h2tg000291l-: Gap from 1660321 to 1731502 on chr2:92516955-93800352\n",
      "h2tg000262l-: Gap from 157396 to 261051 on chr2:94124087-93953719\n",
      "h2tg000019l: Strand not consistent between the first and last segments.\n",
      "h2tg000067l: Strand not consistent between the first and last segments.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr15.hap1.tsv\n",
      "h1tg000397l+: Gap from 171971 to 261090 on chr15:1183245-1192582\n",
      "h1tg000210l-: Gap from 182678 to end on chr15:unknown-1732514\n",
      "h1tg000219l-: Gap from 45913 to end on chr15:unknown-4994382\n",
      "h1tg000552l-: Gap from 282909 to end on chr15:unknown-5533242\n",
      "h1tg000405l+: Gap from 1571440 to 1635604 on chr15:14664182-15143872\n",
      "h1tg000092l-: Gap from 622590 to 803172 on chr15:18299239-17342054\n",
      "h1tg000564l-: Gap from 123926 to end on chr15:unknown-16709032\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr22.hap1.tsv\n",
      "h1tg000334l+: Gap from 488825 to 545924 on chr22:2512608-2548958\n",
      "h1tg000334l+: Gap from 631400 to 713504 on chr22:2606839-2607546\n",
      "h1tg000334l+: Gap from 738413 to 834624 on chr22:2631871-2723310\n",
      "h1tg000334l+: Gap from 917717 to 1141289 on chr22:2821742-3018042\n",
      "h1tg000334l+: Gap from 1153270 to 1239264 on chr22:3028098-3101069\n",
      "h1tg000334l+: Gap from 1317371 to 1734081 on chr22:3188138-3497380\n",
      "h1tg000045l+: Gap from 80113 to 905901 on chr22:8457748-3491072\n",
      "h1tg000045l+: Gap from 1058062 to end on chr22:3594261-unknown\n",
      "h1tg000811l+: Gap from 17064 to end on chr22:4452211-unknown\n",
      "h1tg001135l-: Gap from 18897 to end on chr22:unknown-4816787\n",
      "h1tg004386l-: Gap from 15856 to end on chr22:unknown-6211865\n",
      "h1tg000040l+: Gap from 80859 to end on chr22:8457748-unknown\n",
      "h1tg000219l+: Gap from 260149 to 3270762 on chr22:11284055-10745525\n",
      "h1tg000105l+: Gap from 1076110 to 1127450 on chr22:13798572-13534529\n",
      "h1tg001685l-: Gap from 129216 to end on chr22:unknown-16086095\n",
      "h1tg001723l: Strand not consistent between the first and last segments.\n",
      "h1tg000546l: Strand not consistent between the first and last segments.\n",
      "h1tg000952l: Strand not consistent between the first and last segments.\n",
      "h1tg000197l: Strand not consistent between the first and last segments.\n",
      "h1tg000437l+: Gap from 696407 to end on chr22:23713418-unknown\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr9.hap1.tsv\n",
      "h1tg000160l-: Gap from 241534 to 309963 on chr9:51862465-49766245\n",
      "h1tg000160l: Strand not consistent between the first and last segments.\n",
      "h1tg000208l+: Gap from 310356 to 395900 on chr9:65757079-75323234\n",
      "h1tg000208l: Strand not consistent between the first and last segments.\n",
      "h1tg000208l+: Gap from 533557 to 628276 on chr9:71683392-75310471\n",
      "h1tg000208l: Strand not consistent between the first and last segments.\n",
      "h1tg000208l-: Gap from 1339340 to 1394016 on chr9:57359778-74769416\n",
      "h1tg000208l+: Gap from 1514749 to 1593214 on chr9:74520988-75174421\n",
      "h1tg000208l: Strand not consistent between the first and last segments.\n",
      "h1tg000208l-: Gap from 2331815 to 2383799 on chr9:75467138-74839219\n",
      "h1tg000208l-: Gap from 2578550 to 2700792 on chr9:59629182-71083835\n",
      "h1tg000208l: Strand not consistent between the first and last segments.\n",
      "h1tg001511c+: Gap from 179362 to 251252 on chr9:67241131-73134257\n",
      "h1tg001272l+: Gap from 32895 to end on chr9:76834800-unknown\n",
      "h1tg000277l+: Gap from 53676 to end on chr9:76923734-unknown\n",
      "h1tg000635l: Strand not consistent between the first and last segments.\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr6.hap1.tsv\n",
      "h1tg001539l: Strand not consistent between the first and last segments.\n",
      "h1tg000127l+: Gap from 2453469 to end on chr6:26703056-unknown\n",
      "h1tg000069l+: Gap from 891431 to 1002594 on chr6:32403240-32423311\n",
      "h1tg000374l+: Gap from 242454 to end on chr6:37680962-unknown\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr11.hap2.tsv\n",
      "h2tg000039l+: Gap from 4367295 to 4453310 on chr11:53484692-53587186\n",
      "h2tg000621l-: Gap from 93895 to end on chr11:unknown-53221180\n",
      "h2tg001496l+: Gap from 62873 to end on chr11:91823215-unknown\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr21.hap2.tsv\n",
      "h2tg000436l-: Gap from 191237 to end on chr21:unknown-228814\n",
      "h2tg000043l-: Gap from 33948 to 155070 on chr21:670334-498430\n",
      "h2tg000788l-: Gap from 296874 to end on chr21:unknown-455124\n",
      "h2tg000303l-: Gap from 2197756 to 2263289 on chr21:824832-640434\n",
      "h2tg000303l-: Gap from 2466469 to end on chr21:unknown-455124\n",
      "h2tg000385l+: Gap from 86018 to 347562 on chr21:621793-963063\n",
      "h2tg000285l+: Gap from 138486 to end on chr21:619128-unknown\n",
      "h2tg000666l+: Gap from 63251 to 115745 on chr21:1628496-1687146\n",
      "h2tg000356l+: Gap from 370322 to end on chr21:1902284-unknown\n",
      "h2tg000335l-: Gap from 366151 to 462349 on chr21:6439881-5957871\n",
      "h2tg000148l-: Gap from 2441636 to 4304654 on chr21:10834178-6455185\n",
      "h2tg000148l-: Gap from 4366208 to 4461976 on chr21:6429437-6251998\n",
      "h2tg000148l-: Gap from 4524062 to 4703341 on chr21:6327619-6054000\n",
      "h2tg000148l-: Gap from 4960788 to 5014452 on chr21:6047793-5867066\n",
      "h2tg000148l-: Gap from 5086920 to end on chr21:unknown-5836883\n",
      "h2tg000236l-: Gap from 1395577 to 1555034 on chr21:6840124-6545085\n",
      "h2tg000236l+: Gap from 1699324 to 1761589 on chr21:6680729-6082992\n",
      "h2tg000236l+: Gap from 1773512 to end on chr21:6094650-unknown\n",
      "h2tg000170l+: Gap from 309307 to end on chr21:6803048-unknown\n",
      "h2tg000165l+: Gap from 205858 to end on chr21:10840083-unknown\n",
      "h2tg000058l-: Gap from 10491345 to end on chr21:unknown-10952745\n",
      "h2tg000061l-: Gap from 34167615 to end on chr21:unknown-10962868\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/hap1.ctg.multimap.tsv\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 352, in main\n",
      "    seqid, hapid, tmp = get_scaffold(filename, len_thres=len_thres, gap_thres=gap_thres)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 128, in get_scaffold\n",
      "    is_longctg = check_ctg(ctg, prev, len_thres) # aka pass len_thres and not contained\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 45, in check_ctg\n",
      "    elif ctg[1] < len_thres: return False\n",
      "TypeError: '<' not supported between instances of 'str' and 'int'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr5.hap2.tsv\n",
      "h2tg003305l: Strand not consistent between the first and last segments.\n",
      "h2tg000040l-: Gap from 887466 to 942035 on chr5:49463222-49298854\n",
      "h2tg000073l-: Gap from 12522359 to 12657386 on chr5:49615891-49539938\n",
      "h2tg000073l-: Gap from 13087158 to 13140010 on chr5:47100348-49388728\n",
      "h2tg000073l-: Gap from 13163582 to 13224537 on chr5:49413339-49284625\n",
      "h2tg000346l+: Gap from 157335 to 252480 on chr5:49461599-49438867\n",
      "h2tg000346l+: Gap from 525826 to 674401 on chr5:49471700-47185975\n",
      "h2tg000346l+: Gap from 757757 to 851178 on chr5:49467851-49334728\n",
      "h2tg000346l+: Gap from 1405494 to 1487231 on chr5:49474637-49426643\n",
      "h2tg000346l+: Gap from 1670824 to 1724340 on chr5:49345903-49339962\n",
      "h2tg000204l-: Gap from 127976 to 247742 on chr5:49474758-49339962\n",
      "h2tg000204l-: Gap from 281733 to 353627 on chr5:49374639-49442071\n",
      "h2tg000204l-: Gap from 379360 to 431870 on chr5:49468149-49286487\n",
      "h2tg000204l-: Gap from 532095 to 595325 on chr5:49464721-47225888\n",
      "h2tg000204l-: Gap from 685786 to 845663 on chr5:49485858-49397570\n",
      "h2tg000204l-: Gap from 903191 to end on chr5:unknown-49304167\n",
      "h2tg001574l-: Gap from 116563 to end on chr5:unknown-49406078\n",
      "h2tg000910l-: Gap from 132481 to end on chr5:unknown-49406078\n",
      "h2tg000404l+: Gap from 11681 to 82186 on chr5:49459214-49412867\n",
      "h2tg000404l+: Gap from 241807 to end on chr5:49479062-unknown\n",
      "h2tg000605l+: Gap from 257807 to 340074 on chr5:49471589-49413551\n",
      "h2tg000531l-: Gap from 99428 to 158720 on chr5:49623964-49583650\n",
      "h2tg000052l-: Gap from 157925 to end on chr5:unknown-181890966\n",
      "h2tg000114l-: Gap from 139664 to end on chr5:unknown-181896680\n",
      "Traceback (most recent call last):\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 376, in <module>\n",
      "    main(sys.argv)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 361, in main\n",
      "    find_translocation(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 288, in find_translocation\n",
      "    ctg_segs = stitch_translocation(filename_list, ctgName)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 229, in stitch_translocation\n",
      "    ctg_segs = parse_tsvfile(filename)\n",
      "  File \"/hlilab/yujieguo/K562_analysis/pymodules/ctgSegment2.py\", line 208, in parse_tsvfile\n",
      "    with open(filename,'r') as tsv_file:\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.primary.seqid.tsv'\n",
      "/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/chr12.hap1.tsv\n",
      "h1tg000114l-: Gap from 776233 to 857544 on chr12:35231239-35349849\n",
      "h1tg000001l-: Gap from 19591 to end on chr12:unknown-98715664\n"
     ]
    }
   ],
   "source": [
    "for filename in glob('/hlilab/yujieguo/K562_analysis/r2_chm13_byChr/*.tsv'):\n",
    "    !python ctgSegment2.py -f $filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "10cc1118",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import sys\n",
    "from collections import defaultdict\n",
    "from collections import Counter\n",
    "\n",
    "def parse_ctgpaf(line, ctglen_thres, ctgspan_thres, mq_thres):\n",
    "    '''\n",
    "    Return a list of attributes from a line of paf file. Filter out if the total contig length,\n",
    "    the aligned contig length or mapping quality is shorter than a certain threshold.\n",
    "    ------\n",
    "    Input: \n",
    "        line: one row of a given paf file\n",
    "        ctglen_thres: the total contig length threshold, default is 50kb.\n",
    "        ctgspan_thres: the aligned contig length threshold, default is 20kb.\n",
    "        mq_thres: the mapping quality threshold, default is 1.\n",
    "    Output:\n",
    "        paf_line: a python list containing 10 selected attributes from line.\n",
    "    '''\n",
    "    tmp = line.rstrip().split('\\t')\n",
    "    indices = [*list(range(9)), 11]\n",
    "    # 0ctgName, 1ctgLen, 2ctgStart, 3ctgEnd, 4ctgStrand, 5chrName, 6chrLen, 7chrStart, 8chrEnd, 9MQ\n",
    "    paf_line = [int(tmp[_]) if tmp[_].isdigit() else tmp[_] for _ in indices]\n",
    "    # if alignment is of low quality, filter out\n",
    "    if paf_line[1] < ctglen_thres or (paf_line[3]-paf_line[2]+1)<ctgspan_thres or paf_line[9]<mq_thres or '_' in paf_line[5]:\n",
    "        return None\n",
    "    return paf_line\n",
    "\n",
    "def get_hap(paf_filename):\n",
    "    '''Retrieve hap1 or hap2 from paf file name.'''\n",
    "    tmp = paf_filename.split('.')\n",
    "    hap = ['hap' in i for i in paf_filename.split('.')]\n",
    "    return [tmp[i] for i in range(len(hap)) if hap[i]][0]\n",
    "\n",
    "def sort_ctg_in_chrom(ctg_pafs):\n",
    "    '''sort a list of paf_line of contigs mapped to the same chromosome. \n",
    "    first sort by chrStart, then by ctgStart'''\n",
    "    return sorted(ctg_pafs, key=lambda x: (x[7], x[2]))\n",
    "\n",
    "def split_by_chrom_helper(paf_filename, bed_filename, ctglen_thres, ctgspan_thres, mq_thres):\n",
    "    '''helper to split paf file rows by chromosomes'''\n",
    "    out_dict = defaultdict(list)\n",
    "    ctg_mapcnt_dict = defaultdict(list)\n",
    "    ctg_pafs = []\n",
    "    with open(paf_filename,'r') as paf_file:\n",
    "        for line in paf_file.readlines():\n",
    "            paf_line = parse_ctgpaf(line,ctglen_thres,ctgspan_thres,mq_thres)\n",
    "            if paf_line:\n",
    "                ctg_pafs.append(paf_line)\n",
    "            else:\n",
    "                continue\n",
    "    for ctg_paf in ctg_pafs:\n",
    "        out_dict[ctg_paf[5]].append(ctg_paf)\n",
    "        if ctg_paf[5] not in ctg_mapcnt_dict[ctg_paf[0]]:\n",
    "            ctg_mapcnt_dict[ctg_paf[0]].append(ctg_paf[5])\n",
    "    \n",
    "    rem_list = []\n",
    "    for ctgName, chrList in ctg_mapcnt_dict.items():\n",
    "        if len(chrList) < 2:\n",
    "            rem_list.append(ctgName)\n",
    "    [ctg_mapcnt_dict.pop(key) for key in rem_list]\n",
    "    return out_dict,ctg_mapcnt_dict\n",
    "\n",
    "def split_by_chrom(paf_filename, bed_filename, ctglen_thres = 50000, ctgspan_thres = 10000, mq_thres = 2):\n",
    "    '''\n",
    "    Input:\n",
    "        paf_filename: paf file of contigs aligned to reference genome (e.g, CHM13 or hg38)\n",
    "        bed_filename: \n",
    "    '''\n",
    "    hap = get_hap(paf_filename)\n",
    "    out_dict, ctg_mapcnt_dict = split_by_chrom_helper(paf_filename,ctglen_thres,ctgspan_thres,mq_thres)\n",
    "    \n",
    "    # write each one to files\n",
    "    for chrom, ctg_pafs in out_dict.items():\n",
    "        out_filename = chrom+'.'+hap+'.tsv'\n",
    "        rows = sort_ctg_in_chrom(ctg_pafs)\n",
    "        with open(out_filename, 'w', newline='') as csvfile:\n",
    "            csvwriter = csv.writer(csvfile, delimiter='\\t')\n",
    "            csvwriter.writerows(rows)\n",
    "    \n",
    "    # write contigs that might contain translocation info to one file\n",
    "    with open(f'{hap}.ctg.multimap.tsv','w') as f:\n",
    "        for ctgName, chrList in ctg_mapcnt_dict.items():\n",
    "            for chrName in chrList:\n",
    "                f.write(f'{ctgName}\\t{chrName}\\n')\n",
    "\n",
    "def split_by_chrom_utg(paf_filename, utglen_thres = 20000, utgspan_thres = 5000, mq_thres = 2):\n",
    "    out_dict = split_by_chrom_helper(paf_filename,utglen_thres,utgspan_thres,mq_thres)\n",
    "    # write each one to files\n",
    "    for chrom, ctg_pafs in out_dict.items():\n",
    "        out_filename = f'{chrom}.utg.tsv'\n",
    "        rows = sort_ctg_in_chrom(ctg_pafs)\n",
    "        with open(out_filename, 'w', newline='') as csvfile:\n",
    "            csvwriter = csv.writer(csvfile, delimiter='\\t')\n",
    "            csvwriter.writerows(rows)\n",
    "    return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "47e1a04a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This version tries to assemble chromosomes with translocations.\n",
    "\n",
    "def quick_display(ctg_segs):\n",
    "    '''just get rid of contained ctgs and print out the primary contigs'''\n",
    "    prev = None\n",
    "    output = []\n",
    "    for ctg in ctg_segs:\n",
    "        if iscontained(prev, ctg, threshold=0.05):\n",
    "            continue\n",
    "        else:\n",
    "            output.append(ctg)\n",
    "            prev = ctg\n",
    "    return output\n",
    "\n",
    "def ctgline2ID(ctg):\n",
    "    '''write attributes of a contig to SeqID'''\n",
    "    ctg_len        = f'ctg_len={ctg[1]}'\n",
    "    ctg_range      = f'range={ctg[2]}-{ctg[3]}'\n",
    "    ctg_strand     = f'strand={ctg[4]}'\n",
    "    chr_annotation = f'{ctg[5]}:{ctg[7]}-{ctg[8]}'\n",
    "    return ' | '.join([ctg[0], chr_annotation, ctg_range, ctg_strand, ctg_len])\n",
    "\n",
    "def iscontained(ctg_seg1, ctg_seg2, threshold=0.01):\n",
    "    '''\n",
    "    check if ctg_seg2 is contained in ctg_seg1, the default threshold is 0.01\n",
    "    ***Note: threshold means if less that 1% of the ctg_seg2 base is out of the seg1 range at either end,\n",
    "    it will be counted as contained.\n",
    "    '''\n",
    "    if ctg_seg1 is None:\n",
    "        return False\n",
    "    seg2_len = ctg_seg2[1]\n",
    "    if ctg_seg1[7] <= ctg_seg2[7] and ctg_seg1[8]-ctg_seg2[8] > -seg2_len*threshold:\n",
    "        return True\n",
    "    elif ctg_seg1[8] >= ctg_seg2[8] and ctg_seg2[7]-ctg_seg1[7] > -seg2_len*threshold:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def check_ctg(ctg, prev, len_thres=100000):\n",
    "    '''return False if current contig is shorter than len_thres or is contained in the previous one.\n",
    "    Note: short/contained contig will be saved to HapIDs as it may have important genotype info.'''\n",
    "    # first ctg\n",
    "    if prev is None: return True\n",
    "    # short contig\n",
    "    elif ctg[1] < len_thres: return False\n",
    "    # contained contig\n",
    "#     elif iscontained(prev, ctg, threshold=0.01): \n",
    "#         print('contained contig:')\n",
    "#         print('\\t'.join([str(i) for i in prev]))\n",
    "#         print('\\t'.join([str(i) for i in ctg]))\n",
    "#         return False\n",
    "    else: return True\n",
    "\n",
    "def get_ctg_range(ctg_segs):\n",
    "    first = last = ctg_block = None\n",
    "    if ctg_segs[0][4] != ctg_segs[-1][4]:\n",
    "        print(f'{ctg_segs[0][0]}: Strand not consistent between the first and last segments.')\n",
    "    # get first ctg_seg (smallest ctgstart)\n",
    "    first = ctg_segs[0]\n",
    "    # get last ctg_seg (largest ctgend)\n",
    "    ctg_segs = sorted(ctg_segs, key=lambda l:l[3],reverse=True)\n",
    "    last = ctg_segs[0]\n",
    "    # get the most prominent ctg_seg strand\n",
    "    ctg_segs = sorted(ctg_segs, key=lambda l:l[3]-l[2],reverse=True)\n",
    "    strand = ctg_segs[0][4]\n",
    "    # get smallest chrStart\n",
    "    ctg_segs_chrS = sorted(ctg_segs, key=lambda l:l[7])\n",
    "    chrS = ctg_segs_chrS[0][7]\n",
    "    # get lasrgest chrEnd\n",
    "    ctg_segs_chrE = sorted(ctg_segs, key=lambda l:l[8],reverse=True)\n",
    "    chrE = ctg_segs_chrE[0][8]\n",
    "    \n",
    "    ctgb_len = last[3]-first[2]\n",
    "    ctgb_aln = chrE-chrS\n",
    "    ctgb_len_ratio = round((last[3]-first[2])/first[1],3)\n",
    "    ctgb_aln_ratio = round((chrE-chrS)/first[6],3)\n",
    "    ctg_block = [first[0],first[1],first[2],last[3],strand,first[5],first[6],chrS,chrE,\n",
    "                 ctgb_len,ctgb_aln,ctgb_len_ratio,ctgb_aln_ratio]\n",
    "    return ctg_block\n",
    "\n",
    "def connect_ctg_segs(ctg_segs, gap_thres = 50000):\n",
    "    '''Stitching the contigs with the same name together within the given list of contig segments. \n",
    "    Mark the ones with strand difference and large gap in alignment.'''\n",
    "    # sort by the starting position of contig coordinates\n",
    "    ctg_segs = sorted(ctg_segs, key=lambda l:l[2])\n",
    "    output = []\n",
    "    first = last = None\n",
    "    \n",
    "    # single contig:\n",
    "    if len(ctg_segs) == 1:\n",
    "        output.append(get_ctg_range(ctg_segs))\n",
    "    \n",
    "    # check gaps between contig segments\n",
    "    # segments without large gaps form into blocks\n",
    "    else:\n",
    "        block_idx = 0\n",
    "        for i in range(len(ctg_segs)-1):\n",
    "            prev = ctg_segs[i]\n",
    "            curr = ctg_segs[i+1]\n",
    "            if curr[2]-prev[3] > gap_thres:\n",
    "                print(f'{curr[0]}{curr[4]}: Gap from {prev[3]} to {curr[2]} on {curr[5]}:{prev[8]}-{curr[7]}')\n",
    "                output.append(get_ctg_range(ctg_segs[block_idx:i+1]))\n",
    "                block_idx = i+1\n",
    "        output.append(get_ctg_range(ctg_segs[block_idx:]))\n",
    "    \n",
    "    # check unmapped regions of the contig\n",
    "    last=ctg_segs[-1]\n",
    "    if last[1]-last[3] > gap_thres:\n",
    "        if last[4] == '+':\n",
    "            print(f'{last[0]}{last[4]}: Gap from {last[3]} to end on {last[5]}:{last[8]}-unknown')\n",
    "        else:\n",
    "            print(f'{last[0]}{last[4]}: Gap from {last[3]} to end on {last[5]}:unknown-{last[7]}')\n",
    "    return output\n",
    "\n",
    "def get_scaffold(ctg_byChr_filename, gap_thres=50000, len_thres=50000):\n",
    "    '''Get list of contigs (name+description) that form a scaffold. \n",
    "    Each list element will be the ID in FASTA file.'''\n",
    "    prev = is_longctg = None\n",
    "    # the following three variables are all dictionary of lists\n",
    "    SeqIDs_tmp = defaultdict(list)\n",
    "    SeqIDs = dict()\n",
    "    HapIDs_tmp = defaultdict(list)\n",
    "    \n",
    "    with open(ctg_byChr_filename,'r') as ctg_file:\n",
    "        for line in ctg_file.readlines():\n",
    "            tmp = line.rstrip().split('\\t')\n",
    "            ctg = [int(_) if _.isdigit() else _ for _ in tmp]\n",
    "            is_longctg = check_ctg(ctg, prev, len_thres) # aka pass len_thres and not contained\n",
    "            prev = ctg\n",
    "            if is_longctg:\n",
    "                SeqIDs_tmp[ctg[0]].append(ctg)\n",
    "            else:\n",
    "                HapIDs_tmp[ctg[0]].append(ctg)\n",
    "    for ctg_name, ctg_segs in SeqIDs_tmp.items():\n",
    "        SeqIDs[ctg_name] = connect_ctg_segs(ctg_segs, gap_thres)\n",
    "\n",
    "    return SeqIDs, HapIDs_tmp, SeqIDs_tmp\n",
    "\n",
    "def lprint(lst):\n",
    "    '''A nice way to print list of 1D/2D'''\n",
    "    if isinstance(lst[0],str):\n",
    "        print('\\t'.join([str(val) for val in lst]))\n",
    "    else:\n",
    "        for val in lst:\n",
    "            print('\\t'.join([str(i) for i in val]))\n",
    "    return None\n",
    "\n",
    "def dprint(dicto):\n",
    "    '''A nice way to print dictionary with values of list'''\n",
    "    for keys,values in dicto.items():\n",
    "        if isinstance(values[0],str):\n",
    "            print('\\t'.join([str(val) for val in values]))\n",
    "        else:\n",
    "            for val in values:\n",
    "                print('\\t'.join([str(i) for i in val]))\n",
    "    return None\n",
    "\n",
    "def write2tsv(SeqIDs, chrom, hap, threshold=0.01, len_thres = 50000):\n",
    "    '''\n",
    "    write to two tsv files: \n",
    "    1. only primary contigs\n",
    "    2. full file contained contigs as well in terms of hg38 coordinates\n",
    "    '''\n",
    "    file1_name = f'{chrom}.{hap}.primary.seqid.tsv'\n",
    "    file2_name = f'{chrom}.{hap}.seqid.tsv'\n",
    "    file1 = open(file1_name, 'w', newline='')\n",
    "    file2 = open(file2_name, 'w', newline='')\n",
    "    csvwriter1 = csv.writer(file1, delimiter='\\t')\n",
    "    csvwriter2 = csv.writer(file2, delimiter='\\t')\n",
    "    # write headers of the full file\n",
    "    ctg_header = '\\t'.join(['ctgName','ctgLen','ctgStart','ctgEnd','ctgStrand',\n",
    "                            'chrName','chrLen','chrStart','chrEnd',\n",
    "                            'ctgbLen','chrbLen','ctgbRatio','chrbRatio'])\n",
    "    header = '\\n'.join(['CC\\tDR\\tchr\\tstart\\tend',\n",
    "                       'CC\\tCM\\tp/q-arm',\n",
    "                       f'CC\\tPR\\t{ctg_header}',\n",
    "                       f'CC\\tCO\\t{ctg_header}\\n'])\n",
    "    file1.write(header)\n",
    "    file2.write(header)\n",
    "    \n",
    "    if isinstance(SeqIDs, dict):\n",
    "        ctg_segs = list(chain.from_iterable(SeqIDs.values()))\n",
    "    else:\n",
    "        ctg_segs = SeqIDs\n",
    "    ctg_segs = sorted(ctg_segs, key=lambda l:l[7])\n",
    "    prev = None\n",
    "    for ctg in ctg_segs:\n",
    "        if iscontained(prev, ctg, threshold=threshold):\n",
    "            csvwriter2.writerow(['CO',*ctg])\n",
    "        elif ctg[3]-ctg[2] < len_thres:\n",
    "            continue\n",
    "        else:\n",
    "            csvwriter1.writerow(['PR',*ctg])\n",
    "            file2.write('//\\n')\n",
    "            csvwriter2.writerow(['PR',*ctg])\n",
    "            prev = ctg\n",
    "    file1.close()\n",
    "    file2.close()\n",
    "    return None\n",
    "\n",
    "# begins section of find translocations\n",
    "def parse_tsvfile(filename):\n",
    "    '''The tsvfile starts with 'CC', which indicates the legends for each row.\n",
    "    A row that starts with 'PR' stands for primary contig,\n",
    "    and a row that starts with 'CO' stands for contained contig.'''\n",
    "    global idx\n",
    "    ctg_segs = []\n",
    "    with open(filename,'r') as tsv_file:\n",
    "        for line in tsv_file.readlines():\n",
    "            tsv_line = line.rstrip().split('\\t')\n",
    "            if tsv_line is None:\n",
    "                continue\n",
    "            elif tsv_line[0] == 'CC' and tsv_line[1] == 'PR':\n",
    "                idx = dict()\n",
    "                for i in range(2,len(tsv_line)):\n",
    "                    idx[tsv_line[i]] = i-2\n",
    "            elif tsv_line[0]=='PR' or tsv_line[0]=='CO':\n",
    "                tsv_line = [int(_) if _.isdigit() else _ for _ in tsv_line]\n",
    "                ctg_segs.append(tsv_line[1:])\n",
    "            else:\n",
    "                continue\n",
    "    return ctg_segs\n",
    "\n",
    "def stitch_translocation(filename_list, t_ctgName):\n",
    "    ctg_segs_byChr = dict()\n",
    "    t_idx_byChr = dict()\n",
    "    out_ctg_segs = []\n",
    "    for filename in filename_list:\n",
    "        ctg_segs = parse_tsvfile(filename)\n",
    "        curr_chrName = ctg_segs[0][idx['chrName']]\n",
    "        ctg_segs_byChr[curr_chrName] = ctg_segs\n",
    "    \n",
    "    for chrom, ctg_segs in ctg_segs_byChr.items():\n",
    "        for i in range(len(ctg_segs)):\n",
    "            ctg = ctg_segs[i]\n",
    "            if t_ctgName in ctg:\n",
    "                t_idx_byChr[chrom]=i\n",
    "    t_chroms = list(t_idx_byChr.keys())\n",
    "    for j in range(len(t_chroms)-1):\n",
    "        prev_ctg_segs = ctg_segs_byChr[t_chroms[j]]\n",
    "        curr_ctg_segs = ctg_segs_byChr[t_chroms[j+1]]\n",
    "        prev_t_idx = t_idx_byChr[t_chroms[j]]\n",
    "        prev_t_ctg = prev_ctg_segs[prev_t_idx]\n",
    "        curr_t_idx = t_idx_byChr[t_chroms[j+1]]\n",
    "        curr_t_ctg = curr_ctg_segs[curr_t_idx]\n",
    "        \n",
    "        # this only works if one contig is only connecting one translocation\n",
    "        if prev_t_ctg[idx['ctgStrand']] == curr_t_ctg[idx['ctgStrand']]:\n",
    "            out_ctg_segs.extend(prev_ctg_segs[0:prev_t_idx+1])\n",
    "            out_ctg_segs.extend(curr_ctg_segs[curr_t_idx:])\n",
    "        else:\n",
    "            out_ctg_segs.extend(prev_ctg_segs[0:prev_t_idx+1])\n",
    "            out_ctg_segs.extend(reversed(curr_ctg_segs[0:curr_t_idx+1]))\n",
    "    return out_ctg_segs\n",
    "\n",
    "def stitch_translocation_byPos(filename_list, prev_pos, next_pos):\n",
    "    '''\n",
    "    pos: a tuple of chrName and chrPos\n",
    "    '''\n",
    "    return None\n",
    "\n",
    "def get_filedir(filename):\n",
    "    return '/'.join(filename.split('/')[0:-1])\n",
    "\n",
    "def find_translocation(ctg_multimap_filename):\n",
    "    '''\n",
    "    After the output from get_scaffold is written to two tsv files for each normal chromosome, \n",
    "    we extract the contig names from the hap.ctg.multimap.tsv that indicates possible translocations.\n",
    "    Then we find the corresponding filename_list that contain this ctgName. If not found in primary.seqid.tsv, \n",
    "    then seqid.tsv with contained contig blocks are considered.\n",
    "    '''\n",
    "    filedir = get_filedir(ctg_multimap_filename)\n",
    "    ctg_mapcnt_dict = defaultdict(list)\n",
    "    with open(ctg_multimap_filename,'r') as f:\n",
    "        for line in f.readlines():\n",
    "            tmp = line.rstrip().split('\\t')\n",
    "            ctg_mapcnt_dict[tmp[0]].append(tmp[1])\n",
    "    for ctgName, chrList in ctg_mapcnt_dict.items():\n",
    "        filename_list = []\n",
    "        t_chr_arg = []\n",
    "        if 'h1' in ctgName:\n",
    "            hap = 'hap1'\n",
    "        elif 'h2' in ctgName:\n",
    "            hap = 'hap2'\n",
    "        for chrName in chrList:\n",
    "            filename_list.append(f'{filedir}/{chrName}.{hap}.primary.seqid.tsv')\n",
    "            t_chr_arg.append(chrName.strip('chr'))\n",
    "        ctg_segs = stitch_translocation(filename_list, ctgName)\n",
    "        t_chr_arg = f'{hap}_'.join(t_chr_arg)\n",
    "        t_chr_arg = t_chr_arg+hap\n",
    "        write2tsv(ctg_segs, 'chrt', t_chr_arg)\n",
    "\n",
    "def write2agp(SeqIDs, out_filename):\n",
    "    '''subject to change'''\n",
    "    with open(out_filename, 'w', newline='') as csvfile:\n",
    "        csvwriter = csv.writer(csvfile, delimiter='\\t')\n",
    "        # csvwriter.writerows(SeqIDs.values())\n",
    "        for val in SeqIDs.values():\n",
    "            csvwriter.writerows(val)\n",
    "    return None\n",
    "\n",
    "def write2fasta(fagz_file, SeqIDs, result_file):\n",
    "    ctg_names = SeqIDs.keys()\n",
    "    with gzip.open(fagz_file, \"rt\") as handle:\n",
    "        with open(result_file, \"w\") as f:\n",
    "            for record in SeqIO.parse(handle, \"fasta\"):\n",
    "                if record.id in ctg_names:\n",
    "                    tmp = SeqIDs[record.id]\n",
    "                    # check if the contigs in seqids have different segments\n",
    "                    if isinstance(tmp[0],str):\n",
    "                        record.description = ctgline2ID(tmp)\n",
    "                        SeqIO.write([record], f, \"fasta\")\n",
    "                    else:\n",
    "                        for seqid in tmp:\n",
    "                            seqid_f = ctgline2ID(seqid)\n",
    "                            record.description = seqid_f\n",
    "                            SeqIO.write([record], f, \"fasta\")\n",
    "\n",
    "import sys\n",
    "import os\n",
    "import getopt\n",
    "from glob import glob\n",
    "import csv\n",
    "import gzip\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from itertools import chain\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "30e9ecac",
   "metadata": {},
   "outputs": [],
   "source": [
    "find_translocation('/hlilab/yujieguo/K562_analysis/assembly_try/assem_tsv_files/hap1.ctg.multimap.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71d556cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "paf_filenames = []\n",
    "    len_thres = gap_thres = 50000\n",
    "    for opt, arg in opts:\n",
    "        if opt in ['-f','--paf']:\n",
    "            print(arg)\n",
    "            if '*' in arg:\n",
    "                for filename in glob(arg):\n",
    "                    paf_filenames.append(filename)\n",
    "            else:\n",
    "                paf_filenames.append(arg)\n",
    "        elif opt in ['-l','--len-thres']: len_thres = arg\n",
    "        elif opt in ['-g','--gap-thres']: gap_thres = arg\n",
    "    \n",
    "    if len(paf_filenames) > 0:\n",
    "        for filename in paf_filenames:\n",
    "            # extract chromsome and haplotype info\n",
    "            tmp = filename.split('/')[-1].split('.')\n",
    "            chrom = tmp[0]\n",
    "            hap = tmp[1]\n",
    "            seqid, hapid, tmp = get_scaffold(filename, len_thres=len_thres, gap_thres=gap_thres)\n",
    "            write2tsv(seqid, chrom, hap)\n",
    "    else:\n",
    "        print('Error with PAF file input')\n",
    "        sys.exit(1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
